File: firebase-firestore/src/main/java/com/androidhuman/rxfirebase2/firestore/RxFirebaseFirestore.java
package com.androidhuman.rxfirebase2.firestore;

import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.firestore.SetOptions;
import com.google.firebase.firestore.Source;

import com.androidhuman.rxfirebase2.firestore.model.Value;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import java.util.Map;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

public final class RxFirebaseFirestore {

    @NonNull
    @CheckResult
    public static <T> Single<DocumentReference> add(
            @NonNull CollectionReference ref, @NonNull T value) {
        return new AddDocumentObserver<>(ref, value);
    }

    @NonNull
    @CheckResult
    public static Single<DocumentReference> add(
            @NonNull CollectionReference ref, @NonNull Map<String, Object> value) {
        return new AddMapDocumentObserver(ref, value);
    }

    @NonNull
    @CheckResult
    public static Single<Value<DocumentSnapshot>> data(@NonNull DocumentReference ref) {
        return new DocumentObserver(ref, null);
    }

    @NonNull
    @CheckResult
    public static Single<Value<DocumentSnapshot>> data(
            @NonNull DocumentReference ref, @NonNull Source source) {
        return new DocumentObserver(ref, source);
    }

    @NonNull
    @CheckResult
    public static Single<Value<QuerySnapshot>> data(@NonNull Query query) {
        return new QueryObserver(query);
    }

    @NonNull
    @CheckResult
    public static Single<Value<QuerySnapshot>> data(
            @NonNull Query query, @NonNull Source source) {
        return new QueryObserver(query, source);
    }

    @NonNull
    @CheckResult
    public static Flowable<Value<DocumentSnapshot>> dataChanges(
            @NonNull DocumentReference ref, @NonNull BackpressureStrategy strategy) {
        return dataChanges(ref).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static Observable<Value<DocumentSnapshot>> dataChanges(
            @NonNull DocumentReference ref) {
        return new DocumentChangesObserver(ref);
    }

    @NonNull
    @CheckResult
    public static Flowable<Value<QuerySnapshot>> dataChanges(
            @NonNull Query query, @NonNull BackpressureStrategy strategy) {
        return dataChanges(query).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static Observable<Value<QuerySnapshot>> dataChanges(@NonNull Query query) {
        return new QueryChangesObserver(query);
    }

    @NonNull
    @CheckResult
    public static Completable delete(@NonNull DocumentReference ref) {
        return new DeleteDocumentObserver(ref);
    }

    @NonNull
    @CheckResult
    public static <T> Completable set(@NonNull DocumentReference ref, @NonNull T value) {
        return new SetValueObserver<>(ref, value);
    }

    @NonNull
    @CheckResult
    public static <T> Completable set(
            @NonNull DocumentReference ref, @NonNull T value, @NonNull SetOptions options) {
        return new SetValueObserver<>(ref, value, options);
    }

    @NonNull
    @CheckResult
    public static Completable set(
            @NonNull DocumentReference ref, @NonNull Map<String, Object> value) {
        return new SetMapValueObserver(ref, value);
    }

    @NonNull
    @CheckResult
    public static Completable set(
            @NonNull DocumentReference ref,
            @NonNull Map<String, Object> value, @NonNull SetOptions options) {
        return new SetMapValueObserver(ref, value, options);
    }

    @NonNull
    @CheckResult
    public static Completable update(
            @NonNull DocumentReference ref, @NonNull Map<String, Object> value) {
        return new UpdateMapDocumentObserver(ref, value);
    }

    private RxFirebaseFirestore() {
        throw new AssertionError("No instances");
    }
}

File: firebase-firestore-kotlin/src/main/kotlin/com/androidhuman/rxfirebase2/firestore/RxFirebaseFirestore.kt
@file:Suppress("NOTHING_TO_INLINE", "UNUSED")

package com.androidhuman.rxfirebase2.firestore

import com.androidhuman.rxfirebase2.firestore.model.Value
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.firestore.SetOptions
import com.google.firebase.firestore.Source
import io.reactivex.BackpressureStrategy
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single

inline fun <reified T : Any> CollectionReference.rxAdd(value: T)
        : Single<DocumentReference> = RxFirebaseFirestore.add(this, value)

inline fun CollectionReference.rxAdd(value: Map<String, Any>)
        : Single<DocumentReference> = RxFirebaseFirestore.add(this, value)

inline fun DocumentReference.data()
        : Single<Value<DocumentSnapshot>> = RxFirebaseFirestore.data(this)

inline fun DocumentReference.data(source: Source)
        : Single<Value<DocumentSnapshot>> = RxFirebaseFirestore.data(this, source)

inline fun Query.data()
        : Single<Value<QuerySnapshot>> = RxFirebaseFirestore.data(this)

inline fun Query.data(source: Source)
        : Single<Value<QuerySnapshot>> = RxFirebaseFirestore.data(this, source)

inline fun DocumentReference.dataChanges(strategy: BackpressureStrategy)
        : Flowable<Value<DocumentSnapshot>> = RxFirebaseFirestore.dataChanges(this, strategy)

inline fun DocumentReference.dataChanges()
        : Observable<Value<DocumentSnapshot>> = RxFirebaseFirestore.dataChanges(this)

inline fun Query.dataChanges(strategy: BackpressureStrategy)
        : Flowable<Value<QuerySnapshot>> = RxFirebaseFirestore.dataChanges(this, strategy)

inline fun Query.dataChanges()
        : Observable<Value<QuerySnapshot>> = RxFirebaseFirestore.dataChanges(this)

inline fun DocumentReference.rxDelete()
        : Completable = RxFirebaseFirestore.delete(this)

inline fun <reified T : Any> DocumentReference.rxSet(value: T)
        : Completable = RxFirebaseFirestore.set(this, value)

inline fun <reified T : Any> DocumentReference.rxSet(value: T, options: SetOptions)
        : Completable = RxFirebaseFirestore.set(this, value, options);

inline fun DocumentReference.rxSet(value: Map<String, Any>)
        : Completable = RxFirebaseFirestore.set(this, value)

inline fun DocumentReference.rxSet(value: Map<String, Any>, options: SetOptions)
        : Completable = RxFirebaseFirestore.set(this, value, options)

inline fun DocumentReference.rxUpdate(value: Map<String, Any>)
        : Completable = RxFirebaseFirestore.update(this, value)
